share/wake/lib/core/json.wake

```diff
@@ -538,3 +538,108 @@ export def lastValueInList (values: List JValue): Result JValue Error =
             | flatten
             | deduplicateJObjectKeys lastValueInList
             | rmap JObject
+
+export def jString: JValue => List String = match _
+    JString s = |> s
+    _         = Nil
+
+export def editJString (f: String => String): JValue => JValue = match _
+    JString s = JString (f s)
+    x         = x
+
+export def jInteger: JValue => List Integer = match _
+    JInteger i = |> i
+    _          = Nil
+
+export def editJInteger (f: Integer => Integer): JValue => JValue = match _
+    JInteger i = JInteger (f i)
+    x          = x
+
+export def jBoolean: JValue => List Boolean = match _
+    JBoolean b = |> b
+    _          = Nil
+
+export def editJBoolean (f: Boolean => Boolean): JValue => JValue = match _
+    JBoolean b = JBoolean (f b)
+    x          = x
+
+export def jDouble: JValue => List Double = match _
+    JDouble d = |> d
+    _         = Nil
+
+export def editJDouble (f: Double => Double): JValue => JValue = match _
+    JDouble d = JDouble (f d)
+    x         = x
+
+export def jNumber: JValue => List Double = match _
+    JDouble x  = |> x
+    JInteger x = |> x | mapPartial dint
+    _          = Nil
+
+export def editJNumber (f: Double => Double): JValue => JValue = match _
+    JDouble x  = JDouble (f x)
+    JInteger x = match (dint x)
+        Some dx = JDouble (f dx)
+        None    = JInteger x
+    x          = x
+
+export def jNull: JValue => List Unit = match _
+    JNull = |> Unit
+    _     = Nil
+
+export def editJNull (f: Unit => Unit): JValue => JValue = match _
+    JNull =
+        def _ = f Unit
+        JNull
+    x     = x
+
+export def jObject: JValue => List (List (Pair String JValue)) = match _
+    JObject obj = |> obj
+    _           = Nil
+
+export def editJObject (f: List (Pair String JValue) => List (Pair String JValue)): JValue => JValue = match _
+    JObject obj = JObject (f obj)
+    x           = x
+
+export def jArray: JValue => List (List JValue) = match _
+    JArray arr = |> arr
+    _          = Nil
+
+export def editJArray (f: List JValue => List JValue): JValue => JValue = match _
+    JArray arr = JArray (f arr)
+    x          = x
+
+export def jFlatten: JValue => List JValue = match _
+  JArray a = a
+  _ = Nil
+
+export def editJFlatten: (JValue => JValue) => JValue => JValue = editJArray $ map $ _
+
+export def jAt (i: Integer): JValue => List JValue =
+    jFlatten _ | at i | optionToList
+
+def atE (i: Integer): (a => a) => List a => List a =
+    def edit f l =
+      require Pair first (x, follow) = splitAt i l else l
+      first ++ (f x, follow)
+    edit
+
+export def editJAt (i: Integer): (JValue => JValue) => JValue => JValue =
+    editJArray $ atE i $ _
+
+export def jField (field: String): JValue => List JValue = match _
+    JObject pairs = lookup (_==~field) pairs
+    _             = Nil
+
+export def editJField (field: String): (JValue => JValue) => JValue => JValue =
+    editJObject $ editLookup (_==~field) $ _
```

Comment:
What's the `$ _` for? I'd leave these off from these functions.

Reply:
This is how you gain back composition now that we're using application, you leave a hole at the end (for getters you use `_ |` at the start). the `_` here is the function of type `(JValue => JValue)` being used to edit this field.

Reply:
You're right, or course. It's the `$` that threw me.

If you'd written `_ | editLookup (_==~field) | editJObject` (or even better `editLookup (_==~field) _ | editJObject`) it would have been obvious.

For that matter, I know we can use `$` to keep the edit chains in the same order as the gets, but I wonder if we should just accept a flipped order and use `|`? 

I'm going to push a commit to this PR you can revert if you like, which flips this around.

Reply:
Ok, not changing the order, as discussed on slack.